### PR TITLE
simplify case when literal expression

### PR DESCRIPTION
**Pull request by jinyangli34**
## Description
Simplify case when literal statement if possible.

example:
`case when a = 100 then 1 else 0 end <> 0` should be simplified to `a = 100`

## Additional context and related issues
We have seen many tableau generated queries with unnecessary case when statement. This filter cannot be pushed down to Iceberg, causing more than necessary partitions being processed.

## Release notes

(X) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by mosabua** [comment]
:wave: @jinyangli34  - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by jinyangli34** [comment]
Hi @mosabua, thanks for checking this PR. It will be great to have someone take a look and give some comment/feedbacks.

**Comment by mosabua** [comment]
@martint can you either review or suggest who could review this?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+public class RemoveRedundantCaseWhenLiteral
+        implements Rule<FilterNode>
+{
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public RemoveRedundantCaseWhenLiteral(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
+    {
+        this.plannerContext = plannerContext;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:57]:
```suggestion
        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        this.typeAnalyzer = typeAnalyzer;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:58]:
```suggestion
        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return filter();
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Expression rewritten = ExpressionTreeRewriter.rewriteWith(new Visitor(plannerContext, context.getSession(), context.getSymbolAllocator(), typeAnalyzer), node.getPredicate());
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:70]: static import `rewriteWith`

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        private final Session session;
+        private final SymbolAllocator symbolAllocator;
+        private final TypeAnalyzer typeAnalyzer;
+
+        public Visitor(PlannerContext plannerContext, Session session, SymbolAllocator symbolAllocator, TypeAnalyzer typeAnalyzer)
+        {
+            this.literalInterpreter = new LiteralInterpreter(plannerContext, session);
+            this.session = session;
+            this.symbolAllocator = symbolAllocator;
+            this.typeAnalyzer = typeAnalyzer;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:91]:
```suggestion
            this.literalInterpreter = new LiteralInterpreter(plannerContext, session);
            this.session = requireNonNull(session, "session is null");
            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
            this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        {
+            if ((node.getOperator() == EQUAL || node.getOperator() == NOT_EQUAL)) {
+                if (node.getLeft() instanceof SearchedCaseExpression left &&
+                        node.getRight() instanceof Literal right &&
+                        left.getWhenClauses().stream().map(WhenClause::getResult).allMatch(Literal.class::isInstance) &&
+                        left.getDefaultValue().map(Literal.class::isInstance).orElse(false)) {
+                    Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), node);
+                    Object rightValue = literalInterpreter.evaluate(right, expressionTypes.get(NodeRef.of(right)));
+
+                    Predicate<Expression> matcher = x -> (node.getOperator() == EQUAL) == rightValue.equals(literalInterpreter.evaluate(x, expressionTypes.get(NodeRef.of(x))));
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:105]: nit: rename `x` to `expr`

Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:105]: this comparison is hard to follow (a == b) == c. Can we rewrite it to make it more readable? I.e. extract it to a separate method and cover both EQUAL and NOT_EQUAL cases separately?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+                    List<Expression> positives = left.getWhenClauses().stream()
+                            .filter(wc -> matcher.test(wc.getResult()))
+                            .map(WhenClause::getOperand)
+                            .collect(toImmutableList());
+
+                    List<Expression> negatives;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java:112]: `List<Expression> negatives = ImmutableList.of();`

then drop redundant else below

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java (hunk 5 above)
Comment (jinyangli34) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java]: Created another method to handle EQUAL and NOT_EQUAL separately.

**Comment by mosabua** [comment]
👋 @jinyangli34 could you address the comments from @wendigo to proceed with this PR?

**Comment by jinyangli34** [comment]
> 👋 @jinyangli34 could you address the comments from @wendigo to proceed with this PR?

Yes, I have resolved all the comments. @wendigo can you help take another look?

**Comment by mosabua** [comment]
I think the test failure is unrelated.. can you confirm and review @wendigo and maybe @martint

**Comment by wendigo** [comment]
This now needs a rebase due to AST/IR changes

**Comment by jinyangli34** [comment]
> This now needs a rebase due to AST/IR changes

Hi @wendigo, I made a rebase just now. PTAL

**Review by martint — COMMENTED** [review]
This could be much more easily modeled as pushing down the equality into each of the case branches and then let other optimizations simplify and clean up the expression.

To illustrate:

```
CASE 
    WHEN a = 1 THEN 1
    ELSE 0 
END = 1
```

would turn into:

```
CASE 
    WHEN a = 1 THEN 1 = 1
    ELSE 0 = 1
END
```

which will then simplify to:

```
CASE 
    WHEN a = 1 THEN true
    ELSE false
END
```

Assuming we have an optimization that can turn a CASE statement that evaluates to true in the `then` branch and to false in the `else` branch or viceversa (we may need to implement such optimization), it could be further simplified to:

`a = 1`

What's nice about this approach is that it generalizes to any type and any operator. For instance:

```
CASE 
    WHEN a = 1 THEN 10
    ELSE 1
END > 5
```
=>
```
CASE 
    WHEN a = 1 THEN 10 > 5
    ELSE 1 > 5
END
```
=>
```
CASE 
    WHEN a = 1 THEN true
    ELSE false
END
```
=>
```
a = 1
```

It could even be made to work with function applications:

```
substr(CASE 
    WHEN a = 1 THEN 'hello'
    ELSE 'world'
END, 1, 1) = 'h'
```
=>  
```
CASE 
    WHEN a = 1 THEN substr('hello', 1, 1)
    ELSE substr('world', 1, 1)
END = 'h'
```
=>
```
CASE 
    WHEN a = 1 THEN substr('hello', 1, 1) = 'h'
    ELSE substr('world', 1, 1) = 'h'
END
```
=>
```
CASE 
    WHEN a = 1 THEN true
    ELSE false
END
```

**Comment by jinyangli34** [comment]
I'll do another refactor base on Martint's comment.
